### PR TITLE
Always print lints from plugins, if they're available

### LINKED ISSUE #6122: `cargo clippy -- -W help` runs `clippy`
> `cargo clippy -- -W help` seems to run `clippy`, so it takes a while on a first build and, of course, it requires a project. Would it be possible to print the available lints and quit, similar to `rustc -W help`?

> Context: in `rust-analyzer` we'd like to offer code completions for the various lints, but the only way to get them is to download http://rust-lang.github.io/rust-clippy/master/lints.json.

_Originally posted by @lnicola in https://github.com/rust-lang/rust-clippy/issues/5385#issuecomment-704233618_

Even though the `-Whelp` thing is unstable, it shouldn't have to run Clippy on a crate to print out the help message with all lints.

### PR DESCRIPTION
**Pull request by flip1995**
Currently you can get a list of lints and lint groups by running `rustc
-Whelp`. This prints an additional line at the end:
```
Compiler plugins can provide additional lints and lint groups. To see a listing of these, re-run `rustc -W help` with a crate filename.
```

Clippy is such a "compiler plugin", that provides additional lints.
Running `clippy-driver -Whelp` (`rustc` wrapper) still only prints the
rustc lints with the above message at the end. But when running
`clippy-driver -Whelp main.rs`, where `main.rs` is any rust file, it
also prints Clippy lints. I don't think this is a good approach from a
UX perspective: Why is a random file necessary to print a help message?

This PR changes this behavior: Whenever a compiler callback
registers lints, it is assumed that these lints come from a plugin and
are printed without having to specify a Rust source file.

Fixes rust-lang/rust-clippy#6122

cc @Manishearth @ebroto for the Clippy changes.

### CODE REVIEW

**Comment by Manishearth** [comment]
r+ on making this happen, I'll let eddy review the PR (though i should be able to if i get time)

**Review by ebroto — APPROVED** [review]
Clippy part LGTM :+1:

**Comment by flip1995** [comment]
@eddyb @Manishearth This PR is 1 1/2 months old and waiting for review. This would be nice to have from a Clippy perspective.

**Comment by lnicola** [comment]
For more context, it's also useful for `rust-analyzer` because it can use it to offer completions for `#[allow]` and `#[deny]` attributes.

**Comment by flip1995** [comment]
> For more context, it's also useful for rust-analyzer because it can use it to offer completions for #[allow] and #[deny] attributes.

Oh right, that's why I did this in the first place. I already forgot lol :smile:

File: compiler/rustc_driver/src/lib.rs
Comment (Manishearth) [inline compiler/rustc_driver/src/lib.rs:964]: let's actually just get rid of this? It's not relevant to clippy, and compiler plugins are not a common thing.

File: src/tools/clippy/src/lintlist/mod.rs
Comment (Manishearth) [inline src/tools/clippy/src/lintlist/mod.rs:1]: YAY

**Review by Manishearth — APPROVED** [review]

File: compiler/rustc_driver/src/lib.rs
Comment (flip1995) [reply compiler/rustc_driver/src/lib.rs]: I reworded it in e54c060. Removing this would also involve dealing with this case in the match statement somehow. I think printing a message, that tools can provide more lints is a good way to deal with this case.

**Pull request merged**
